### PR TITLE
Add lane components

### PR DESCRIPTION
**Pull request by aniafafe**
Resolve: #81

### CODE REVIEW

File: Project/Gem/CMakeLists.txt [hunk 1]
```diff
@@ -89,6 +89,9 @@ ly_add_target(
 target_depends_on_ros2_package(${gem_name}.Private.Object  vision_msgs REQUIRED)
 target_depends_on_ros2_package(${gem_name} vision_msgs REQUIRED)
 
+target_depends_on_ros2_package(${gem_name}.Private.Object lane_provider_msgs REQUIRED)
```
Comment (adamdbrw) [inline Project/Gem/CMakeLists.txt:92]: add to the same target depends (it supports a list of deps)

File: Project/Gem/Source/Navigation/LanesService.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#include <AzCore/Component/Component.h>
+#include <AzCore/Component/EntityId.h>
+#include <AzCore/std/containers/vector.h>
+
+namespace ROS2::Demo
+{
+    using ListTracksRequest = std::shared_ptr<lane_provider_msgs::srv::ListTracks::Request>;
+    using ListTracksResponse = std::shared_ptr<lane_provider_msgs::srv::ListTracks::Response>;
+
+    class LanesService : public AZ::Component
```
Comment (adamdbrw) [inline Project/Gem/Source/Navigation/LanesService.h:23]: Rename to LaneServiceComponent

File: Project/Gem/Source/Navigation/LaneComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    void LaneComponent::Activate()
+    {
+        AZ::TickBus::Handler::BusConnect();
+    }
+
+    void LaneComponent::OnTick(float deltaTime, AZ::ScriptTimePoint time)
+    {
+        for (auto const path : m_paths)
```
Comment (adamdbrw) [inline Project/Gem/Source/Navigation/LaneComponent.cpp:56]:
```suggestion
        for (auto const& path : m_paths)
```

File: Project/Gem/Source/Navigation/LaneComponent.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            }
+
+            auto address = splinePtr->GetAddressByFraction(fraction);
+            const AZ::Vector3 position_ = splinePtr->GetPosition(address);
+            const AZ::Vector3 tangent_ = splinePtr->GetTangent(address);
+            const AZ::Vector3 normal_ = splinePtr->GetNormal(address);
+
+            const AZ::Matrix3x3 rot_ = AZ::Matrix3x3::CreateFromColumns(tangent_, normal_, tangent_.Cross(normal_));
+            const AZ::Transform goalTransform_ = AZ::Transform::CreateFromMatrix3x3AndTranslation(rot_, position_);
+            auto m_idealGoal_ = splineTransform * goalTransform_;
```
Comment (adamdbrw) [inline Project/Gem/Source/Navigation/LaneComponent.cpp:95]: Please adjust naming conventions. (with name_ and m_name)

File: Project/Gem/Source/Navigation/LaneComponent.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            pose.pose.orientation.w = rotation.GetW();
+
+            path.poses.push_back(pose);
+        }
+
+        return path;
+    }
+
+    void LaneComponent::Deactivate()
+    {
```
Comment (adamdbrw) [inline Project/Gem/Source/Navigation/LaneComponent.cpp:121]: Also disconnect from Bus

File: Project/Gem/Source/Navigation/LaneComponent.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    AZStd::string LaneComponent::GetLaneName()
+    {
+        return GetEntity()->GetName();
+    }
+
+    lane_provider_msgs::msg::LanePaths LaneComponent::GetLanePathMsgs()
```
Comment (adamdbrw) [inline Project/Gem/Source/Navigation/LaneComponent.cpp:129]: Does not change between Activate-Deactivate. Cache in onTick

File: Project/Gem/Source/Navigation/LanesService.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                ec->Class<LanesService>("LanesService", "LanesService")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2::Demo")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC_CE("Game"))
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default,
+                        &LanesService::m_lanesEntities,
+                        "Set of lanes in the warehouse",
+                        "Set of lanes in the warehouse")
```
Comment (adamdbrw) [inline Project/Gem/Source/Navigation/LanesService.cpp:42]: Nit: remove the specific context not essential to design

File: Project/Gem/Source/Navigation/LanesServiceComponent.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        AZStd::string requestedLaneName(request->lane_name.c_str());
+
+        if (requestedLaneName == "")
+        {
+            for (const auto entityId : m_tracks)
+            {
+                AZ::Entity* entity = nullptr;
+                AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationRequests::FindEntity, entityId);
+                LaneComponent* laneComponent = entity->FindComponent<LaneComponent>();
```
Comment (michalpelka) [inline Project/Gem/Source/Navigation/LanesServiceComponent.cpp:75]: Add here assertion

File: Project/Gem/Source/Navigation/LaneComponent.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+    void LaneComponent::OnTick(float deltaTime, AZ::ScriptTimePoint time)
+    {
+        m_lanePathMsgs.lane_name = GetLaneName().c_str();
+
+        for (auto const& path : m_paths)
+        {
+            m_pathsMsgs[path.m_orderNumber] = CalculatePoses(path);
+            AZ::Entity* pathEntity = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(pathEntity, &AZ::ComponentApplicationRequests::FindEntity, path.m_entityId);
```
Comment (michalpelka) [inline Project/Gem/Source/Navigation/LaneComponent.cpp:62]: Add assertion here as well

Comment (michalpelka) [reply Project/Gem/Source/Navigation/LaneComponent.cpp:62]: Consider early return/continue and warning.

**Review by michalpelka — APPROVED** [review]

**Pull request merged**